Semaphore V4

cedoor opened the pull request:
<!-- Please refer to our contributing documentation for any questions on submitting a pull request -->
<!--- Provide a general summary of your changes in the Title above -->

## Description

<!--- Describe your changes in detail -->

This PR updates the Semaphore codebase based on the findings from the analysis and research period described in the [research repository](https://github.com/semaphore-protocol/research/).

The only supported tree depth is 10 at the moment. Another PR (based on issue https://github.com/semaphore-protocol/semaphore/issues/482) will be created to support a tree depth range from 1 to 32.

This is not the final version of the PR. The tasks that will follow are listed below. However, an initial review is necessary before proceeding.

**ZK circuits**

- The identity trapdoor and identity nullifier have been replaced with a single value, the private key, from which an EdDSA public key is derived. The Poseidon hash of the public key is the new identity commitment.
- Some input and output names in the circuits have been changed.
  - External Nullifier → Scope
  - SignalHash → Message
  - NullifierHash → Nullifier
  - Siblings → Tree Siblings
  - Path Indices → Tree Indices
- The new circuit uses [Anonymous Components](https://docs.circom.io/circom-language/anonymous-components-and-tuples/#anonymous-components) to make the code clearer and more concise (24 lines of code without comments, rather than 90).
- The circuit used to calculate the Merkle tree root is now part of an external package ([`@zk-kit/circuits`](https://github.com/privacy-scaling-explorations/zk-kit/tree/main/packages/circuits) - [`BinaryMerkleRoot`](https://github.com/privacy-scaling-explorations/zk-kit/blob/main/packages/circuits/circom/binary-merkle-root.circom)) and supports dynamic depths.
- [Circomkit](https://github.com/erhant/circomkit) has been used for tests and scripts
- The new circuit will be published on NPM (`@semaphore-protocol/circuit`)

**JS libraries**

* The core Semaphore libraries have been updated to support the new [LeanIMT](https://github.com/privacy-scaling-explorations/zk-kit/blob/main/packages/imt/src/lean-imt/lean-imt.ts) data structure and the new V4 circuits.
* The identity library now supports the generation of public/private keys using EdDSA and Baby Jubjub, as well as signatures.
* The group and proof libraries have been greatly simplified.
* The proof library now automatically downloads `wasm`/`zkey` artifacts on browsers and also on NodeJS.
* JS tests have been moved outside the `src` folder. A new `tests` folder has been created for them.
* `@semaphore-protocol/heyauthn` tests have been updated to support the new core libraries
* `@semaphore-protocol/hardhat` has been updated to support the new contracts

**Solidity contracts**

* The contracts have been updated with the new [LeanIMT](https://github.com/privacy-scaling-explorations/zk-kit/blob/main/packages/imt.sol/contracts/internal/InternalLeanIMT.sol).
* `SemaphoreGroups.sol` has been adapted to the new data structure, which does not require zero nodes and has a static depth.
* `SemaphoreGroups.sol` now also manages access to functions (group admins).
* The verifier has been updated with the latest version of the template provided by [SnarkJS](https://github.com/iden3/snarkjs/blob/master/templates/verifier_groth16.sol.ejs).
* The `VerifiedProof` event now includes all the parameters of the proof, allowing anyone to verify the proof off-chain (https://github.com/semaphore-protocol/semaphore/issues/328).

## Related Issue

<!--- This project accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

Closes #357 

## Does this introduce a breaking change?

-   [x] Yes
-   [ ] No

<!-- If this introduces a breaking change, please describe the impact and migration path for existing applications below. -->

## Other information

### Next tasks

Before merging this PR to the `main` branch it is necessary to work on other tasks, which will be merged to this branch first.

- https://github.com/semaphore-protocol/semaphore/issues/530
- https://github.com/semaphore-protocol/semaphore/issues/531
- https://github.com/semaphore-protocol/semaphore/issues/532
- https://github.com/semaphore-protocol/semaphore/issues/533
- https://github.com/semaphore-protocol/semaphore/issues/486
- https://github.com/semaphore-protocol/semaphore/issues/534
- https://github.com/semaphore-protocol/semaphore/issues/352
- https://github.com/semaphore-protocol/semaphore/issues/482
- https://github.com/semaphore-protocol/semaphore/issues/329
- https://github.com/semaphore-protocol/semaphore/issues/485
- https://github.com/semaphore-protocol/semaphore/issues/521
- https://github.com/semaphore-protocol/semaphore/issues/522
- https://github.com/semaphore-protocol/semaphore/issues/524
- https://github.com/semaphore-protocol/semaphore/issues/543
- https://github.com/semaphore-protocol/semaphore/issues/544
- https://github.com/semaphore-protocol/semaphore/issues/545

cedoor committed:
feat(circuits): update semaphore circuits

re #357

cedoor committed:
ci: add circuit tests to workflows

re #357

cedoor committed:
chore(circuits): update circom configuration

re #357

cedoor committed:
feat(identity): update identity with eddsa

cedoor committed:
refactor(identity): add identity commitment attribute

cedoor committed:
feat(group): update group with lean-imt

cedoor committed:
refactor(identity): update commitment variable name

cedoor committed:
feat(proof): update proof package with new circuit

cedoor committed:
chore: update env variables

cedoor committed:
refactor(contracts): remove extension contracts

cedoor committed:
feat(contracts): update contracts with new imt

cedoor committed:
refactor: update verifiers

cedoor committed:
chore(proof): update zk-kit groth16 pkg version

cedoor committed:
fix(contracts): reverse proof b points

cedoor committed:
refactor: hash message and scope

cedoor committed:
refactor(contracts): move access logic to group contract

cedoor committed:
refactor(heyauthn): update tests and doc

cedoor committed:
refactor(hardhat): update hardhat pkg tasks

cedoor committed:
chore: move test files outside src

cedoor committed:
refactor(contracts): fix eslint warnings

cedoor committed:
docs: update readme files

cedoor committed:
chore(proof): create env bundles

cedoor committed:
refactor(circuits): update variable name

vplasencia commented on packages/proof/src/generate-proof.ts:
```diff
@@ -0,0 +1,74 @@
+    snarkArtifacts?: SnarkArtifacts
+): Promise<SemaphoreProof> {
+    const leafIndex = group.indexOf(identity.commitment)
+    const merkeProof = group.generateMerkleProof(leafIndex)
```
```suggestion
    const merkleProof = group.generateMerkleProof(leafIndex)
```

cedoor committed:
fix(proof): set correct var name

cedoor committed:
refactor(circuits): update merkle proof variable names

numtel commented on packages/contracts/contracts/Semaphore.sol:
```diff
@@ -150,15 +104,15 @@ contract Semaphore is ISemaphore, SemaphoreGroups {
         uint256[8] calldata proof
-    ) external override {
+    ) external override onlyExistingGroup(groupId) {
         uint256 merkleTreeDepth = getMerkleTreeDepth(groupId);
```
This merkle tree depth is 0 when there's only 1 group member. The merkle root is simply the identity commitment at that time.

This just needs to be changed to `getMerkleTreeSize` instead and it will  work as intended.

See the [definition in the zk-kit imt.sol repo](https://github.com/privacy-scaling-explorations/zk-kit/blob/main/packages/imt.sol/contracts/internal/InternalLeanIMT.sol#L8)

0xjei replied:
Thank you @numtel. Please, see #535 to reflect your requested changes

0xjei commented on .github/workflows/production.yml:
```diff
@@ -63,7 +63,14 @@ jobs:
               uses: actions/setup-node@v1
               with:
-                  node-version: 16.x
+                  node-version: "18"
```
Should we switch to latest LTS version of Node?

v20 ( LTS )	8 months ago (18 Apr 2023) - 20.10.0 (22 Nov 2023)

cedoor replied:
I think it makes sense as v20 is LTS and then stable enough.

0xjei replied:
See [#536](https://github.com/semaphore-protocol/semaphore/pull/536)

0xjei commented on packages/circuits/LICENSE:
```diff
@@ -0,0 +1,21 @@
+MIT License
+
+Copyright (c) 2023 Ethereum Foundation
```
nit: move to 2024?

cedoor replied:
Yes we should update all the licenses in V4

0xjei replied:
See [#537](https://github.com/semaphore-protocol/semaphore/pull/537)

0xjei commented on packages/circuits/semaphore.circom:
```diff
@@ -1,90 +1,25 @@
-    signal output out;
-
-    component poseidon = Poseidon(1);
+    merkleRoot <== BinaryMerkleRoot(MAX_DEPTH)(identityCommitment, merkleProofLength, merkleProofIndices, merkleProofSiblings);
```
neat +1

0xjei commented on packages/contracts/contracts/interfaces/ISemaphore.sol:
```diff
@@ -3,26 +3,20 @@ pragma solidity 0.8.4;
     error Semaphore__MerkleTreeDepthIsNotSupported();
     error Semaphore__MerkleTreeRootIsExpired();
     error Semaphore__MerkleTreeRootIsNotPartOfTheGroup();
     error Semaphore__YouAreUsingTheSameNillifierTwice();
```
```suggestion
    error Semaphore__YouAreUsingTheSameNullifierTwice();
```

cedoor replied:
Good catch, the other occurrences must also be updated.

0xjei replied:
See [#538](https://github.com/semaphore-protocol/semaphore/pull/538)

0xjei commented on packages/identity/README.md:
```diff
@@ -49,8 +49,8 @@
 
-| This library provides a class that can be used to create identities compatible with the Semaphore [circuits](https://github.com/semaphore-protocol/semaphore/tree/main/circuits). Each identity contains two secret values: _trapdoor_ and _nullifier_, and one public value: _commitment_. The Poseidon hash of the secret values is the identity secret, and its hash is the identity commitment. |
-| --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| This library provides a class that can be used to create identities compatible with the Semaphore [circuits](https://github.com/semaphore-protocol/semaphore/tree/main/circuits). Each identity contains an EdDSA private key, its public key, and the identity commitment, which is the Poseidon hash of the public key. |
```
nit: maybe is worth to put an external link to EdDSA and Poseidon?

btw, definitely easier to read and understand!

cedoor replied:
Yes, it makes sense 👍🏽

0xjei replied:
See [#539](https://github.com/semaphore-protocol/semaphore/pull/539)

0xjei commented on packages/proof/README.md:
```diff
@@ -86,18 +87,22 @@ import { utils } from "ethers"
-})
+const fullProof1 = await generateProof(identity, group, message, scope)
+
+// You can also specify the maximum shaft depth supported by the proof.
```
nit

```suggestion
// You can also specify the maximum tree depth supported by the proof.
```

0xjei commented on packages/proof/src/get-snark-artifacts.browser.ts:
```diff
@@ -0,0 +1,9 @@
+
+export default async function getSnarkArtifacts(treeDepth: number): Promise<SnarkArtifacts> {
+    return {
+        wasmFilePath: `https://semaphore.cedoor.dev/artifacts/${treeDepth}/semaphore.wasm`,
```
nit: switch the host from your personal @cedoor to PSE one?

cedoor replied:
This is temporary. We'll change host as soon as we run the ceremony and have the actual artifacts.

0xjei reviewed (changes requested):
Outstanding job :1st_place_medal: - Thank you @cedoor.

Left some comments and minor nits around.

cedoor committed:
Update packages/proof/README.md

Co-authored-by: Giacomo <giacomo.corrias7@gmail.com>

cedoor committed:
chore(contracts): update solidity and hardhat versions

cedoor committed:
chore(contracts): removed unused snarkjs templates

cedoor committed:
refactor(contracts): update script to verify contracts

cedoor committed:
chore(contracts): downgrade prettier & solhint packages

cedoor committed:
test(proof): add longer timeout for generiting proofs

0xjei committed:
fix: use merkleTreeSize instead of merkleTreeDepth to support groups with one member

0xjei committed:
build: move github actions to V4 along with new NodeJS LTS version (20)

0xjei committed:
refactor: update npm types for node up to latest NodeJS LTS version (20)

0xjei committed:
chore: update copyright year

0xjei committed:
fix: typo for ISemaphore contract interface custom error

0xjei committed:
docs: add missing external references for EdDSA and Poseidon

cedoor committed:
chore(contracts): update solidity version

cedoor committed:
chore(circuits): udpate zk-kit dependency

cedoor committed:
feat(contracts): add new function to verify proofs without nullifiers

cedoor committed:
ci: remove unnecessary cache actions from workflows

vplasencia commented on packages/circuits/semaphore.circom:
```diff
@@ -1,90 +1,25 @@
-    signal input identityNullifier;
-    signal input identityTrapdoor;
+template Semaphore(MAX_DEPTH) {
+    signal input privateKey;
```
What about using the name `secret` or `secretScalar` instead of `privateKey` because it can be confusing with the private key of the identity, and they are not the same.

cedoor replied:
Yes, it makes sense!

cedoor replied:
Please see: https://github.com/semaphore-protocol/semaphore/pull/480/commits/076e5de03c3646590e33f830dd75905e6fc0103d

vplasencia reviewed (approved):
Great work! ✨

vplasencia committed:
feat(contracts): add many members using new event

vplasencia committed:
chore(contracts): update zk-kit imt version

vplasencia committed:
chore(contracts): update imt version

vplasencia committed:
test(contracts): update add members test

cedoor committed:
refactor(circuits): rename privateKey with secret

vplasencia committed:
refactor(contracts): optimize gas

vplasencia committed:
test(contracts): remove skip tests

vplasencia committed:
docs(contracts): add code docs for the members added event

cedoor committed:
refactor(contracts): add 1 verifier for each tree depth

cedoor committed:
refactor(contracts): update tests, tasks and scripts

cedoor committed:
chore(contracts): remove base contracts

cedoor committed:
refactor(proof): support multi tree depths (1-12)

cedoor committed:
style(contracts): format code with prettier

cedoor committed:
chore: remove goerli from supported networks

re #352

cedoor committed:
style(subgraph): format code with prettier

cedoor committed:
chore: improve scripts to clean repo

cedoor committed:
chore(contracts): add open-zeppelin deployer

cedoor committed:
chore: update env file

cedoor committed:
perf(contracts): merge verifiers

cedoor committed:
chore(contracts): deploy contracts on sepolia

cedoor committed:
chore(subgraph): update subgraph with new contract events

re #543

cedoor committed:
style(contracts): format code with prettier

cedoor committed:
style(subgraph): format code with prettier

cedoor committed:
refactor(contracts): improve task to deploy

cedoor committed:
chore(subgraph): update docker compose file

cedoor committed:
revert(contracts): re-add script to verify contracts

cedoor committed:
refactor(subgraph): set correct type for tree depth

cedoor committed:
chore: update cli and data packages

cedoor committed:
refactor(subgraph): update merkle tree var name

cedoor committed:
fix: adjust cli and data packages

cedoor committed:
chore(cli): update contracts-hardhat template

cedoor committed:
chore(hardhat): update hardhat package dependencies

cedoor committed:
refactor(proof): include merkle depth in the proof

cedoor committed:
chore(contracts): update proof pkg var names

cedoor committed:
refactor(identity): convert priv key buffer to bigint

cedoor committed:
chore(cli): update cli monorepo templates

cedoor committed:
test(data): temporary comment tests for sepolia testnets

cedoor committed:
style: format code with prettier

cedoor committed:
style(cli): update cli templates' contract artifacts

cedoor committed:
test(data): update ethers tests

cedoor committed:
chore: v4.0.0-alpha

cedoor committed:
chore: remove stableVersion auto-generated field

cedoor committed:
chore: remove stableVersion auto-generated field

cedoor committed:
refactor(contracts): create SemaphoreProof struct

cedoor committed:
refactor: rename proof with points

cedoor committed:
style: format code with prettier

cedoor committed:
refactor(proof): remove folder creation before downloading artifacts

re #582

cedoor committed:
chore(contracts): create script to add mock data on testnets

cedoor committed:
refactor(proof): add merkle proof parameter

cedoor committed:
chore(proof): replace zk-kit groth16 with snarkjs

swellander committed:
chore(cli): change `create` command message

converts all `npm` commands to `yarn` commands

re #592

cedoor committed:
chore: v4.0.0-alpha.1

cedoor committed:
chore: remove stableVersion auto-generated field

cedoor committed:
fix(cli): add check for groups with 1 member

cedoor committed:
feat(contracts): add function to get group admin

re #574

cedoor committed:
build(identity): config rollup to handle browser bundle

re #481

cedoor committed:
docs: update shields license labels

re #466

cedoor committed:
build: fix bundle paths

cedoor committed:
chore: v4.0.0-alpha.2

cedoor committed:
chore: remove stableVersion auto-generated field

cedoor committed:
build(identity): include poseidon code in bundle

cedoor committed:
chore: v4.0.0-alpha.3

cedoor committed:
chore: remove stableVersion auto-generated field

cedoor committed:
fix(data): update semaphore contract abi

cedoor committed:
chore: v4.0.0-alpha.4

cedoor committed:
chore: remove stableVersion auto-generated field

cedoor committed:
chore(circuits): add circuit scheme

cedoor committed:
chore(circuits): update circuit scheme

cedoor committed:
chore(circuits): update circuit scheme

cedoor committed:
refactor(contracts): update scripts to save deployed contract addresses

cedoor committed:
style(contracts): format code with prettier

cedoor committed:
chore(cli): sync monorepo templates with boilerplate

cedoor committed:
chore(cli): replace localhost with 127.0.0.1

cedoor committed:
chore(cli): update cli monorepo template app descriptions

cedoor committed:
chore: remove all fullProof occurrences

cedoor committed:
feat(core): create core package

cedoor committed:
chore(cli): use semaphore core package

cedoor committed:
chore: update ts, rollup dependencies

cedoor committed:
Update packages/identity/package.json

Co-authored-by: Vivian Plasencia <v.pcalana@gmail.com>

cedoor committed:
build: fix default exports

cedoor committed:
chore: v4.0.0-alpha.5

cedoor committed:
chore: remove stableVersion auto-generated field

cedoor committed:
fix(core): add full import path in esm modules

cedoor committed:
chore: v4.0.0-alpha.6

cedoor committed:
chore: remove stableVersion auto-generated field

cedoor committed:
refactor(group): empty group root return 0

cedoor committed:
feat(proof): support message/scope as strings

cedoor committed:
docs(proof): update usage sections

cedoor committed:
refactor(proof): create function to convert to bigint

cedoor committed:
Update CONTRIBUTING.md

cedoor committed:
fix(data): update ethers dependency

cedoor committed:
feat(group): add export/import methods

cedoor committed:
Merge pull request #625 from semaphore-protocol/fix/sepolia-testnets

Update `ethers` dependency

cedoor committed:
Merge pull request #626 from semaphore-protocol/feat/export-group

New Group `export`/`import` methods

cedoor committed:
chore: v4.0.0-alpha.7

cedoor committed:
chore: remove stableVersion auto-generated field

cedoor committed:
fix(data): update provider parameters

cedoor committed:
chore: upgrade jest to v29

re #633

cedoor committed:
chore(group): remove console.log

cedoor committed:
chore(data): move test file to tests folder

cedoor committed:
refactor: use ethers conditional exports

re #632

cedoor committed:
refactor(proof): create 0-dependency download script

re #635

cedoor committed:
chore(circuits): update imt dependency

cedoor committed:
Merge pull request #630 from semaphore-protocol/fix/ethers-providers

New `ethers``s provider parameters

cedoor committed:
Merge pull request #634 from semaphore-protocol/refactor/conditional-exports

 `ethers` conditional exports

cedoor committed:
Merge pull request #636 from semaphore-protocol/refactor/download-artifacts

0-dependency script to download SNARK artifacts

cedoor committed:
chore: v4.0.0-alpha.8

cedoor committed:
chore: remove stableVersion auto-generated field

cedoor committed:
feat(utils): create new utils package

cedoor committed:
feat(utils): add functions to check object types

cedoor committed:
feat(utils): add functions to check bigint types

cedoor committed:
refactor(proof): add functions to check types

cedoor committed:
build(utils): set correct file extensions

cedoor committed:
chore: update lockfile

cedoor committed:
Merge pull request #643 from semaphore-protocol/feat/utils-pkg

New `utils` package with functions two check types

cedoor committed:
docs(docs): update v4 alpha docs to v4 beta

vplasencia committed:
refactor(cli): update cli code and dependencies

vplasencia committed:
refactor(cli-template-contracts-hardhat): update cli contracts hardhat template

vplasencia committed:
refactor(cli-template-contracts-hardhat): update cli contracts hardhat template

vplasencia committed:
style(cli-template-contracts-hardhat): add spaces to the end of files

vplasencia committed:
style(cli-template-contracts-hardhat): add spaces to the end of files

vplasencia committed:
refactor(cli-template-monorepo-ethers): update cli monorepo ethers template

vplasencia committed:
style(cli-template-monorepo-ethers): format code with prettier

vplasencia committed:
style(cli-template-monorepo-ethers): add a space to the end of files

vplasencia committed:
chore: update yarn lock file

vplasencia committed:
refactor(cli-template-monorepo-subgraph): update the cli monorepo subgraph template

vplasencia committed:
style(cli-template-monorepo-subgraph): format code with prettier

vplasencia committed:
style(cli-template-monorepo-subgraph): add space to the end of files

cedoor committed:
docs(docs): update V3 github links

cedoor committed:
docs(docs): update V4-beta docs links

cedoor committed:
chore(docs): set V4-beta as default version

cedoor committed:
docs(docs): update boilerplate links

cedoor committed:
Merge pull request #650 from semaphore-protocol/ref/cli

Update CLI code and dependencies

vplasencia committed:
refactor(proof): integrate pack unpack groth16 proof

cedoor committed:
Merge pull request #657 from semaphore-protocol/ref/pack-unpack-proof

Integrate `packGroth16Proof` and `unpackGroth16Proof` into Semaphore v4

zkfriendly committed:
feat(data): add `getGroupMembers` to `SemaphoreSubgraph`

zkfriendly committed:
feat(data): add `getGroupValidatedProofs` to `SemaphoreSubgraph`

zkfriendly committed:
refactor(data): remove unnecessary `?? []`

cedoor committed:
Merge pull request #660 from zkfriendly/feat/semaphore-v4

Adds `getGroupMembers` & `getGroupValidatedProofs` functions to `SemaphoreSubgraph`

cedoor committed:
Merge pull request #655 from semaphore-protocol/docs/v4-beta

New V4-beta documentation

cedoor committed:
docs(circuits): add documentation to circuits

This commit add more documentation to the Semaphore circuits, with the aim of providing an overview
of the code used and to refer to other resources for futher information.

re #483

vplasencia committed:
refactor(cli-template-contracts-hardhat): update configs

vplasencia committed:
refactor(cli-template-contracts-hardhat): update configs

vplasencia committed:
refactor(cli-templates-monorepo-ethers): update configs

vplasencia committed:
refactor(cli-template-monorepo-subgraph): update configs

vplasencia committed:
refactor(cli-template-monorepo-subgraph): remove unused env example file

vplasencia committed:
refactor(cli-template-monorepo-subgraph): update variable name

vplasencia committed:
refactor(cli-template-monorepo-ethers): update eslintignore

vplasencia committed:
refactor(cli-template-monorepo-ethers): update gitignore files

vplasencia committed:
refactor(cli-templates-monorepo-subgraph): update gitignore files

vplasencia committed:
Merge branch 'ref/cli-template-contracts-hardhat' of https://github.com/semaphore-protocol/semaphore into ref/cli-template-contracts-hardhat

vplasencia committed:
refactor: update yarn lock file

vplasencia committed:
refactor(cli-template-contracts-hardhat): update tsconfig file

cedoor committed:
Merge pull request #651 from semaphore-protocol/ref/cli-template-contracts-hardhat

Update `cli-template-contracts-hardhat`

vplasencia committed:
refactor(cli-template-monorepo-ethers): fix semaphore dependencies

vplasencia committed:
refactor(cli-template-monorepo-subgraph): fix semaphore version

vplasencia committed:
refactor(cli-template-monorepo-ethers): fix semaphore version in tests

vplasencia committed:
refactor(cli-template-contracts-hardhat): update tsconfig in contracts

vplasencia committed:
refactor(cli-template-monorepo-subgraph): update tsconfig contracts

vplasencia committed:
refactor(cli-template-contracts-hardhat): fix semaphore version in tests

vplasencia committed:
Merge branch 'feat/semaphore-v4' into ref/cli-template-monorepo-ethers

vplasencia committed:
Merge branch 'feat/semaphore-v4' into ref/cli-template-monorepo-subgraph

cedoor committed:
Merge pull request #653 from semaphore-protocol/ref/cli-template-monorepo-subgraph

Update `cli-template-monorepo-subgraph`

cedoor committed:
chore: add npm script to test specific libs

cedoor committed:
docs(identity): add more documentation to code

The commit adds more documentation to the Identity class and update the TypeDoc link. It also
removes the class method to verify signatures as the static method can be used instead.

re #483

cedoor committed:
Merge pull request #652 from semaphore-protocol/ref/cli-template-monorepo-ethers

Update `cli-template-monorepo-ethers`

cedoor committed:
docs(identity): add more documentation to code

The commit adds more documentation to the proof package functions and updates the TypeDoc link.

re #483

cedoor committed:
chore: update lockfile

cedoor committed:
docs(circuits): add more documentation to secret signal

re #483

cedoor committed:
docs(identity): add links to zk-kit types

re #483

cedoor committed:
chore: update lockfile

cedoor committed:
docs(circuits): add link to identity package

re #483

cedoor committed:
docs(circuits): revert order of hash pre-images on comments

re #483

cedoor committed:
Merge pull request #663 from semaphore-protocol/docs/identity

More documentation on Identity lib code

cedoor committed:
docs(identity): fix jsdoc link syntax

cedoor committed:
Update packages/proof/src/generate-proof.ts

Co-authored-by: Vivian Plasencia <v.pcalana@gmail.com>

cedoor committed:
Update packages/circuits/semaphore.circom

Co-authored-by: Vivian Plasencia <v.pcalana@gmail.com>

cedoor committed:
docs(circuits): minor improvements on doc comments

re #483

cedoor committed:
Merge pull request #661 from semaphore-protocol/docs/circuits

More documentation on Circom code

cedoor committed:
Merge pull request #664 from semaphore-protocol/docs/proof

More documentation on proof lib code

cedoor committed:
docs(group): add more documentation to code

The commit adds more documentation to the group package functions and updates the TypeDoc link.

re #483

cedoor committed:
docs: fix typedoc links

cedoor committed:
style(data): format code with prettier

cedoor committed:
chore: upgrade yarn to v4

re #662

cedoor committed:
chore(contracts): update openzeppelin dev dependency version

re #662

cedoor committed:
Merge pull request #667 from semaphore-protocol/chore/yarn-v4

Upgrade Yarn to v4

vplasencia committed:
docs(contracts): add code comments to contracts

vplasencia committed:
style(contracts): style code with prettier

vplasencia committed:
docs(contracts): update code comments

vplasencia committed:
fix(contracts): remove comment line that was causing an error when compiling contracts

cedoor committed:
Merge pull request #669 from semaphore-protocol/docs/contracts

Add code comments to contracts

cedoor committed:
Merge pull request #665 from semaphore-protocol/docs/group

More documentation on group lib code

cedoor committed:
Merge pull request #666 from semaphore-protocol/docs/fix-typedoc-links

Correct TypeDoc links

vplasencia committed:
docs(contracts): remove code comment

cedoor committed:
Merge pull request #668 from semaphore-protocol/chore/new-openzeppelin-version

New OpenZeppelin version
